### ISSUE TITLE
Offline database/database encryption cannot be opened on Android

### ISSUE THREAD

**Issue by kibhub** [tutao/tutanota#4560]
**Description**
Users are reporting an error message when trying to do offline log in. The following error message appear and prevents offline log in:

`net.sqlcipher.database.SQLiteDiskIOException`

After the first appearance of the message it persist at every login (both offline and online).

Some users report that dismissing the message while online, enables them to log in but we also have reports of failed log in attempts.

Deleting the stored credentials solves the problem temporarily but the message returns at some point.

**Attempts to reproduce**
We haven't been able to reproduce this issue. The following has been tested on Android 10 without any luck so far:

- fresh install > store credentials > log out > go offline > do offline login
- fresh install > store credentials > change encryption mode to Biometrics only > go offline > do offline login
- fresh install > store credentials > change encryption mode to system password > go offline > do offline login
- multiple attempts at logins and logouts both while offline and online with and without multiple accounts stored in the app.

**Details**
 - OS: Andoid 10, 11, 12
 - Version Tutanota v3.98.24

**Stacktrace**
```
net.sqlcipher.database.SQLiteDiskIOException: error code 10: disk I/O error
at net.sqlcipher.database.SQLiteStatement.native_execute(Native Method)
at net.sqlcipher.database.SQLiteStatement.execute(SQLiteStatement.java:61)
at net.sqlcipher.database.SQLiteDatabase.execSQL(SQLiteDatabase.java:2493)
at de.tutao.tutanota.offline.AndroidSqlCipherFacade.run(AndroidSqlCipherFacade.kt:52)
at de.tutao.tutanota.ipc.SqlCipherFacadeReceiveDispatcher.dispatch(SqlCipherFacadeReceiveDispatcher.kt:41)
at de.tutao.tutanota.ipc.AndroidGlobalDispatcher.dispatch(AndroidGlobalDispatcher.kt:39)
at de.tutao.tutanota.RemoteBridge.invokeMethod(RemoteBridge.kt:153)
at de.tutao.tutanota.RemoteBridge.handleMessageFromWeb(RemoteBridge.kt:90)
at de.tutao.tutanota.RemoteBridge$initMessageChannel$1$onMessage$1.invokeSuspend(RemoteBridge.kt:58)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

## Test notes
All on Android:
 - [ ] New app install, can log in
 - [ ] Deleting credentials
 - [ ] Switching between accounts

**Comment by JSkier21** [comment]
On Android 12, Oneplus 8T, logging out and back in, and it seems to go away. I think perhaps this was just the wrong time after an upgrade to the app? I also tried to get the error back by doing the steps to reproduce, and it hasn't come back since logging out and back in. Likewise, I will let you know if it comes back. 

On Android 11, OnePlus Nord N10, the error has yet to occur.

**Comment by gkalausis** [comment]
> On Android 12, Oneplus 8T, logging out and back in, and it seems to go away. I think perhaps this was just the wrong time after an upgrade to the app? I also tried to get the error back by doing the steps to reproduce, and it hasn't come back since logging out and back in. Likewise, I will let you know if it comes back. 
> 
> On Android 11, OnePlus Nord N10, the error has yet to occur. 
> 

Same issue for me for Xperia 5ii android 12 and to log out and then login solved the issue.

**Comment by JSkier21** [comment]
It started coming back a few hours later for me on OP 8T, unfortunately. I have sent in the error reports.

**Comment by ZettaPloom** [comment]
> It started coming back a few hours later for me on OP 8T, unfortunately. I have sent in the error reports.

Same for me on Xiaomi Poco X3 Pro, deleting stored credentials and logging in again solved the issue for a few hours but the error persisted

**Comment by charlag** [comment]
This happened to my myself over the weekend, deleting credentials helped for now.
I think it might be related to the double login issue when you click on a notification, did anyone observe the same?

**Comment by charlag** [comment]
We've investigated and we've found how to fix it and also have a theory on what is causing it.

The version of android-sqlite-sqlcipher that we build differs in size from the one from the developers. When we build it locally we get a version with the same size and it also works without issues. We noticed that our Jenkins job was also not pointing to a master branch so we are not sure how exactly the last artifact was built. Replacing the .aar with new version produces working releases even for existing databases.

The underlying issue itself can be one of the following:
 - Some kind of a timing issue. The database access is not synchronized in Android code and maybe library size difference makes a difference in concurrency. This is supported by the fact that sometimes deleting the database helps (how big the database is influences the timing).
 - Something in the build environment was different.

**Comment by charlag** [comment]
Update: there was a testing issue, it still does not seem to work with our build

**Comment by charlag** [comment]
The issue is resolved now (#4577).
The actual issue was with the build flags for the library. It is likely that by default SQLite tried to create a temp file and it didn't work.
We've also strengthened concurrency handling just in case.
See https://github.com/sqlcipher/android-database-sqlcipher/issues/600
